tests/blackbox/testlib/infra.py

```diff
@@ -121,3 +125,37 @@ def __exit__(self, exception_type, exception_value, traceback):
             if exception_value is None:
                 raise rexc
             raise rexc from exception_value
+
+
+def test_permissions(dbus_method, *args, permissions):
+    """
+    Permissions: True if should fail with dropped permissions,
+    False if should succeed with dropped permissions.
+    """
+    _ROOT = 0
+    _NON_ROOT = 1
+    _PERMISSIONS = False
+
+    euid = os.geteuid()
+    if euid != _ROOT:
+        raise RuntimeError(
+            "The current process should be running with a euid of 0, but the current euid is %d."
+            % euid
+        )
+    dbus_method(*args)
+    StratisDbus.close_connection()
+
+    os.seteuid(_NON_ROOT)
+    try:
+        dbus_method(*args)
+    except dbus.exceptions.DBusException as exception:
+        if exception.get_dbus_name() == "org.freedesktop.DBus.Error.AccessDenied":
+            _PERMISSIONS = True
+        else:
+            raise exception
+    except Exception as exception:
+        os.seteuid(_ROOT)
+        raise exception
```

Comment:
It looks like you've kind of combined the approaches here. You'd probably want to drop `except Exception` or else not check the type and just match against `dbus.freedesktop.DBus.Error.AccessDenied` in the first `except` block. Currently with this code, any D-Bus error that is not access denied will leave the testing process as non-root.